src/bsp/src/stm32/src/InterlocManager.cpp

```diff
@@ -66,3 +82,174 @@ bool InterlocManager::constructUWBHeader(uint16_t destinationId,
+
+    int i = 0;
+    double val = 0;
+    while (device.getState() == DW_STATE::RESPOND_CALIB) { // find exit condition
+        val += receiveTWRSequence(destinationId, device);
+        i++;
+        if (i > NB_CALIB_TRIES - 1) {
+
+            i = 0;
+            // Compute the distance error between acquired and actual
+            error = (int32_t)((val * 100 / NB_CALIB_TRIES) - m_distanceForCalibCm);
+            val = 0;
+            // convert distance error to tick count in DW device
+            // P type control, P  = 0.9
+            // TODO could had a PID type control #futureImprovement
+            dwCountOffset = 0.9 * error * DW_INTERNAL_CLOCK_RFEQ / SPEED_OF_LIGHT / 100;
+            // antenna delay is considered equal on Rx and Tx
+            device.setTxAntennaDLY((dwCountOffset >> 1) + device.getTxAntennaDLY());
+            device.setRxAntennaDLY((dwCountOffset >> 1) + device.getRxAntennaDLY());
+            if (dwCountOffset >> 1 <= 1) {
+                device.setState(DW_STATE::CALIBRATED);
+                m_calibFinishedCallback(m_calibFinishedCallbackContext);
+            }
+        }
+        Task::delay(50);
+    }
+}
+
+double InterlocManager::receiveTWRSequence(uint16_t destinationId, Decawave& device) {
+    UWBRxFrame rxFrame;
+    uint64_t pollRxTs;
+    uint64_t respTxTs;
+    uint64_t finalRxTs;
+    UWBMessages::TWRResponse responseMsg{};
+    double varA;
+    double varB;
+    double varC;
+    double varD;
+    double distanceEval;
+    double tof;
+    uint64_t tofDtu;
+
+    // receive poll message
+    device.receive(rxFrame, 0);
+
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_POLL) {
+        return 0;
+    }
+
+    pollRxTs = rxFrame.m_rxTimestamp;
+
+    // construct and send Response message
+    constructUWBHeader(destinationId, UWBMessages::DATA, UWBMessages::TWR_RESPONSE,
+                       (uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    respTxTs = (pollRxTs + (POLL_RX_TO_RESP_TX_DLY_UUS * UUS_TO_DWT_TIME))>>8;
+
+    device.transmit((uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    device.transmitDelayed((uint8_t*)&responseMsg,sizeof(responseMsg),respTxTime);
+
+    // wait for Final message reception
+    device.receive(rxFrame, 0);
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_FINAL) {
+        return 0;
+    }
+
+    UWBMessages::TWRFinal* finalTWRFrame = (UWBMessages::TWRFinal*)(rxFrame.m_frame);
+    device.getTxTimestamp(&respTxTs);
+    finalRxTs = rxFrame.m_rxTimestamp;
+
+    // evaluate distance
+    varA = (double)finalTWRFrame->m_respMinPoll;
+    varB = (double)((uint32_t)finalRxTs - (uint32_t)respTxTs);
+    varC = (double)(finalTWRFrame->m_finaleMinResp);
+    varD = (double)((uint32_t)respTxTs - (uint32_t)pollRxTs);
```

Comment:
why cast them to uint32_t? couldn't we loose information? If not, why are they uint64_t to begin with?

Reply:
I removed the cast,
yes we loose information, but that's the point, it prevents breaking the system if the  clock wraps mid communication